node/rpc-client/src/client.rs

```diff
@@ -1,42 +1,188 @@
 // Copyright 2020 ChainSafe Systems
 // SPDX-License-Identifier: Apache-2.0, MIT
 
-#![allow(clippy::all)]
-#![allow(unused_variables, dead_code)]
+use jsonrpc_v2::{Error, Id, RequestObject};
+use log::{error, info};
+use regex::Regex;
+use serde::de::DeserializeOwned;
+use serde::Deserialize;
+use serde_json::Value;
+use std::{env, fmt};
 
-use blocks::{header::json::BlockHeaderJson, tipset_json::TipsetJson};
-use cid::json::CidJson;
-use jsonrpsee::raw::RawClient;
-use jsonrpsee::transport::http::HttpTransportClient;
-use message::unsigned_message::json::UnsignedMessageJson;
+const DEFAULT_MULTIADDRESS: &str = "/ip4/127.0.0.1/tcp/1234/http";
+const DEFAULT_URL: &str = "http://127.0.0.1:1234/rpc/v0";
+const API_INFO_KEY: &str = "FULLNODE_API_INFO";
+const RPC_ENDPOINT: &str = "rpc/v0";
 
-jsonrpsee::rpc_api! {
-    pub Filecoin {
-        /// Auth
-        #[rpc(method = "Filecoin.AuthNew", positional_params)]
-        fn auth_new(perm: Vec<String>) -> String;
-        /// Chain
-        #[rpc(method = "Filecoin.ChainGetBlock", positional_params)]
-        fn chain_get_block(cid: CidJson) -> BlockHeaderJson;
+/// Error object in a response
+#[derive(Deserialize)]
+struct JsonRpcError {
+    pub code: i64,
+    pub message: String,
+}
+
+impl fmt::Display for JsonRpcError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Error ({}): {}", self.code, self.message)
+    }
+}
+
+#[derive(Deserialize)]
+struct JsonRpcResponse<T> {
+    pub jsonrpc: String,
+    pub result: Option<T>,
+    pub error: Option<JsonRpcError>,
+    pub id: Option<Id>,
+}
+
+/// Parses an ip4 multiaddress into an HTTP URL
+fn multiaddress_to_url(ma_str: String) -> String {
+    // Example haystack: "/ip4/127.0.0.1/tcp/1234/http"
+    let regex = Regex::new(r"/ip4/(?P<protocol>.*)/tcp/(?P<host>.*)/(?P<port>.*)").unwrap();
```

Comment:
Dont you need to check if its http? This regex i think only captures ip4 and tcp

Reply:
Good point, it should work with the output from `lotus auth api-info`. I'll be working on that now in my multiaddress parsing refactor.

Reply:
This works now:

`FULLNODE_API_INFO="/ip4/127.0.0.1/tcp/1234/http" lotus auth api-info --perm admin`

> FULLNODE_API_INFO=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJBbGxvdyI6WyJyZWFkIiwid3JpdGUiLCJzaWduIiwiYWRtaW4iXX0.6dMalXxBGadVxVLzbxAanYV--7ihmIGK3eLrtjhEawQ:/ip4/127.0.0.1/tcp/1234/http

`FULLNODE_API_INFO=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJBbGxvdyI6WyJyZWFkIiwid3JpdGUiLCJzaWduIiwiYWRtaW4iXX0.6dMalXxBGadVxVLzbxAanYV--7ihmIGK3eLrtjhEawQ:/ip4/127.0.0.1/tcp/1234/http target/debug/forest auth create-token -p write`

> 2021-05-28T16:25:29.142Z INFO  rpc_client::client > Using JSON-RPC v2 HTTP URL: http://127.0.0.1:1234/rpc/v0
> eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJBbGxvdyI6WyJyZWFkIiwid3JpdGUiXX0.KBtaQC9K8Tnu2tbyyf0pv0MMD8-kbA7cCVATLqS6F4o